snekbox/memfs.py

```diff
@@ -0,0 +1,137 @@
+
+from snekbox.snekio import FileAttachment
+
+log = logging.getLogger(__name__)
+
+PID = os.getpid()
+
+NAMESPACE_DIR = Path("/memfs")
+NAMESPACE_DIR.mkdir(exist_ok=True)
+
+
+def mount_tmpfs(name: str, size: int | str) -> Path:
+    """Create and mount a tmpfs directory."""
+    tmp = NAMESPACE_DIR / name
+    tmp.mkdir()
+    # Mount the tmpfs
+    subprocess.check_call(
+        [
+            "mount",
+            "-t",
+            "tmpfs",
+            "-o",
+            f"size={size}",
+            "tmpfs",
+            str(tmp),
+        ]
+    )
+    return tmp
+
+
+def unmount_tmpfs(name: str) -> None:
+    """Unmount and remove a tmpfs directory."""
+    tmp = NAMESPACE_DIR / name
+    subprocess.check_call(["umount", str(tmp)])
+    # Unmounting will not remove the original folder, so do that here
+    tmp.rmdir()
+
+
+class MemFS:
+    """A temporary directory using tmpfs."""
+
+    assignment_lock = BoundedSemaphore(1)
+    assigned_names: set[str] = set()  # Pool of tempdir names in use
+
+    def __init__(self, instance_size: int) -> None:
+        """
+        Create a temporary directory using tmpfs.
+
+        size: Size limit of each tmpfs instance in bytes
+        """
+        self.path: Path | None = None
+        self.instance_size = instance_size
+
+    @property
+    def name(self) -> str | None:
+        """Name of the temp dir."""
+        return self.path.name if self.path else None
+
+    @property
+    def home(self) -> Path | None:
+        """Path to home directory."""
+        return Path(self.path, "home") if self.path else None
+
+    def __enter__(self) -> MemFS:
+        """Mounts a new tempfs, returns self."""
+        with self.assignment_lock:
+            for _ in range(10):
+                # Combine PID to avoid collisions with multiple snekbox processes
+                if (name := f"{PID}-{uuid4()}") not in self.assigned_names:
+                    self.path = mount_tmpfs(name, self.instance_size)
+                    self.assigned_names.add(name)
+                    break
+            else:
+                raise RuntimeError("Failed to generate a unique tempdir name in 10 attempts")
+
+        self.mkdir("home")
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        self.cleanup()
+
+    def mkdir(self, path: Path | str, chmod: int = 0o777) -> Path:
+        """Create a directory in the tempdir."""
+        folder = Path(self.path, path)
+        folder.mkdir(parents=True, exist_ok=True)
+        folder.chmod(chmod)
+        return folder
+
+    def attachments(
+        self, max_count: int, max_size: int | None = None
+    ) -> Generator[FileAttachment, None, None]:
+        """Return a list of attachments in the tempdir."""
+        count = 0
+        # Look for any file starting with `output`
+        for file in self.home.glob("output*"):
```

Comment:
Thoughts on requiring writing to an "output" directory rather than requiring file names to be prefixed with `output`?

Reply:
Hm, I think that sounds reasonable. I was originally trying to avoid creating any other directories in the instance /home since now there's a potential FileExistsError if an user provides a file called `output` in the request. Though I suppose we could check for that as a validation?

Reply:
Implemented since [78b4b6a](https://github.com/python-discord/snekbox/pull/159/commits/78b4b6af18a40db3162aad56eb726a26c5a74e8c)